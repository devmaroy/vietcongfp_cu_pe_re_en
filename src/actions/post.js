// Post Actions

import database from '../firebase/firebase';
import history from '../routers/history';
import { 
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    POST_LOADING,
    CLEAR_POST_LOADING,
    COMMENT_LOADING,
    CLEAR_COMMENT_LOADING,
    GET_ERRORS
} from './types';


// Get Posts
export const startGetPosts = () => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( 'posts' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const posts = [];

                snapshot.forEach( ( childSnapshot ) => {
                    posts.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch({
                    type: GET_POSTS,
                    posts
                });
    
                dispatch( clearPostLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Post
export const startGetPost = ( id ) => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( `posts/${ id }` )
            .once( 'value' )
            .then( ( snapshot ) => {
                if ( snapshot.exists() ) {
                    const post = {
                        id: snapshot.key,
                        ...snapshot.val()
                    };
        
                    dispatch({
                        type: GET_POST,
                        post
                    });
    
                    dispatch( clearPostLoading() );
                } else {
                    // Simple redirect to non existent route
                    history.push( '/not-found' );
                }
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Post
export const startAddPost = ( postData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'posts' )
            .push( postData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_POST,
                    post: {
                        id: ref.key,
                        ...postData 
                    }
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
} ;


// Edit Post 
export const startEditPost = ( id, updates ) => {
    return ( dispatch ) => {
        return database
            .ref( `posts/${ id }` )
            .update( updates )
            .then( ( ref ) => {
                dispatch({
                    type: EDIT_POST,
                    id,
                    updates
                });
    
                history.push( `/posts/${ id }` );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Post
export const startRemovePost = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `posts/${ id }` )
            .remove()
            .then( () => {
                dispatch({
                    type: REMOVE_POST,
                    id
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Comments By Post Id
export const startGetCommentsByPostId = ( postId ) => {
    return ( dispatch ) => {
        dispatch( setCommentLoading() );

        return database
            .ref( 'comments' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const comments = [];
            
                snapshot.forEach( ( childSnapshot ) => {
                    if ( childSnapshot.val().postId === postId ) {
                        comments.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    }
                });
    
                dispatch({
                    type: GET_COMMENTS,
                    comments
                });
    
                dispatch( clearCommentLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Comment
export const startAddComment = ( commentData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'comments' ) 
            .push( commentData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_COMMENT,
                    comment: {
                        id: ref.key,
                        ...commentData
                    }
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Comment
export const startRemoveComment = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `comments/${ id }` )
            .remove()
            .then(() => {
                dispatch({
                    type: REMOVE_COMMENT,
                    id
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Post Loading
export const setPostLoading = () => ({
    type: POST_LOADING
});


// Clear Post Loading
export const clearPostLoading = () => ({
    type: CLEAR_POST_LOADING
});


// Comment Loading
export const setCommentLoading = () => ({
    type: COMMENT_LOADING
});


// Clear Comment Loading
export const clearCommentLoading = () => ({
    type: CLEAR_COMMENT_LOADING
});