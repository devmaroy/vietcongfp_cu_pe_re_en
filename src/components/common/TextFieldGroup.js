import React from 'react';
import PropTypes from 'prop-types';

const TextFieldGroup = ( { name, placeholder, value, error, info, type, onChange, disabled, className } ) => {
    return (
        <div>
            <input
                type={ type }
                placeholder={ placeholder }
                name={ name }
                value={ value }
                onChange={ onChange }
                disabled={ disabled }
                className={ className }
            />
            { info && <small>{ info }</small> }
            { error && <div>{ error }</div> }
        </div>
    );
};

TextFieldGroup.propTypes = {
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    value: PropTypes.string.isRequired,
    info: PropTypes.string,
    error: PropTypes.string,
    type: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
    className: PropTypes.string
};
  
TextFieldGroup.defaultProps = {
    type: 'text'
};

export default TextFieldGroup;