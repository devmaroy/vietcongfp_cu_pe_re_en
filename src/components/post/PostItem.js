import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import moment from 'moment';
import { Link } from 'react-router-dom';
//{ id, title, body, name, createdAt, showActions, isAuthenticated }
const PostItem = ( props ) => {
    const { post, isAuthenticated, showActions } = props;
  
    return (
        <li>
            <Link to={ `/posts/${ post.id }` }>
                <h3>{ post.title }</h3>
            </Link>
            <p>{ post.body }</p>
            <h4>Author: { post.name }</h4>
            <h5>Date: { moment( post.createdAt ).format('MMMM Do YYYY, h:mm:ss a') }</h5>

            { isAuthenticated && (
                <Link to={ `/posts/${ post.id }/edit` }>edit</Link>
            )}

            { showActions && (
                <Link to={ `/posts/${ post.id }` }>Read more</Link>
            )}
        </li>
    );
};

PostItem.propTypes = {
    post: PropTypes.object.isRequired
}

PostItem.defaultProps = {
    showActions: true
};

const mapStateToProps = ( state ) => ({
    isAuthenticated: state.auth.isAuthenticated
});

export default connect( mapStateToProps )( PostItem );