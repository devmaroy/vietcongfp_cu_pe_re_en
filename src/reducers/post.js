// Post Reducer

import {
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    SET_POST_LOADING,
    CLEAR_POST_LOADING,
    SET_COMMENT_LOADING,
    CLEAR_COMMENT_LOADING
} from '../actions/types';

const postReducerDefaultState = {
    posts: [],
    post: {},
    comments: [],
    isPostLoading: false,
    isCommentLoading: false
};

export default ( state = postReducerDefaultState, action ) => {
    switch ( action.type ) {
        
        case GET_POSTS:
            return {
                ...state,
                posts: action.posts.reverse()
            };

        case GET_POST:
            return {
                ...state,
                post: action.post
            };

        case ADD_POST:
            return {
                ...state,
                posts: [ action.post, ...state.posts ]
            };

        case EDIT_POST:
            return {
                ...state,
                posts: state.posts.map( ( post ) => {
                    if ( post.id === action.id ) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };

        case REMOVE_POST:
            return {
                ...state,
                posts: state.posts.filter( ( post ) => post.id !== action.id )
            };
        
        case GET_COMMENTS: 
            return {
                ...state,
                comments: action.comments.reverse()
            }
            
        case ADD_COMMENT:
            return {
                ...state,
                comments: [ action.comment, ...state.comments ]
            };

        case REMOVE_COMMENT:
            return {
                ...state,
                comments: state.comments.filter( ( comment ) => comment.id !== action.id )
            };

        case SET_POST_LOADING:
            return {
                ...state,
                isPostLoading: true
            };

        case SET_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: true
            };

        case CLEAR_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: false
            };

        case CLEAR_POST_LOADING:
            return {
                ...state,
                isPostLoading: false
            };
    
        default:
            return state;
    };
};