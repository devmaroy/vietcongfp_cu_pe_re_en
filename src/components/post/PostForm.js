import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextFieldGroup from '../common/TextFieldGroup';
import TextAreaFieldGroup from '../common/TextAreaFieldGroup';
import Spinner from '../common/Spinner';

class PostForm extends React.Component {
    state = {
        title: this.props.post ? this.props.post.title : '',
        body: this.props.post ? this.props.post.body : '',
        createdAt: this.props.post ? this.props.post.createdAt : Date.now(),
        errors: {}
    };

    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Title
        if ( ! this.state.title ) {
            formIsValid = false;
            errors[ 'title' ] = 'Title cannot be empty';
        }

        // Body
        if ( ! this.state.body ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text cannot be empty';
        } else if ( this.state.body.length < 15 ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text must be at least 15 characters.'
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;
    
        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const postData = {
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt,
                name: this.props.user.name
            };
            
            this.props.onSubmit( postData );
        }
    }

    render() {
        const { errors } = this.state;

        return (
            <form onSubmit={ this.onSubmit }>
                <TextFieldGroup 
                    name="title"
                    placeholder="Title"
                    value={ this.state.title }
                    onChange={ this.handleChange }
                    error={ errors.title }
                />
                <TextAreaFieldGroup
                    name="body"
                    placeholder="Text"
                    value={ this.state.body }
                    onChange={ this.handleChange }
                    error={ errors.body }
                />
                <button>Save Post</button>
            </form>
        );
    };
};

PostForm.propTypes = {
    user: PropTypes.object.isRequired,
    post: PropTypes.object,
};

const mapStateToProps = ( state ) => ({
    user: state.auth.user
});

export default connect( mapStateToProps )( PostForm );