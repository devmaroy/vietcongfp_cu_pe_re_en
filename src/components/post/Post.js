import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPostById } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostItem from './PostItem';
import CommentPage from '../comment/CommentPage';

export class Post extends React.Component {
    componentDidMount() {
        this.props.startGetPostById( this.props.match.params.id );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postContent = <Spinner />;       
        } else {
            postContent = (
                <div>
                    <PostItem key={ post.id } post={ post } showActions={ false } isSingle={ true } />
                    <CommentPage postId={ post.id } />
                </div>
            )
        }

        return (
            <div className="content-container">
                <div className="page">
                    { postContent }
                </div>
            </div>
        );
    }
}

Post.propTypes = {
    startGetPostById: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPostById: ( id ) => dispatch( startGetPostById( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( Post );     