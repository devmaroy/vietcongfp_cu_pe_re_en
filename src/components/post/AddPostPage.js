import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startAddPost } from '../../actions/post';
import PostForm from './PostForm';

export class AddPostPage extends React.Component {
    onSubmit = ( postData ) => {
        this.props.startAddPost( postData );
    };

    render() {
        return (
            <div className="content-container">
                <div className="page">
                    <h2>Create Post</h2>
                    <PostForm 
                        onSubmit={ this.onSubmit }
                    />
                </div>
            </div>
        )
    };
};

AddPostPage.propTypes = {
    startAddPost: PropTypes.func.isRequired
};

const mapDispatchToProps = ( dispatch ) => ({
    startAddPost: ( postData ) => dispatch( startAddPost( postData ) )
});

export default connect( undefined, mapDispatchToProps )( AddPostPage );