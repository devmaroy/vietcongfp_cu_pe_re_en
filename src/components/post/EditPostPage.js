import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPost, startEditPost, startRemovePost } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostForm from './PostForm';

class EditPostPage extends React.Component {
    componentDidMount() {
        this.props.startGetPost( this.props.match.params.id );
    };

    onRemove = () => {
        this.props.startRemovePost( this.props.match.params.id );
    };

    onSubmit = ( updates ) => {
        this.props.startEditPost( this.props.match.params.id, updates );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postEditContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postEditContent = <Spinner />
        } else {
           postEditContent = (
              <div>
                    <h3>Edit Post</h3>
                    <PostForm
                        onSubmit={ this.onSubmit }
                        post={ post }
                    />
                    <button onClick={ this.onRemove }>Remove</button>
              </div>
           );
        }

        return (
            <div>
                { postEditContent }
            </div>
        );
    }
}

EditPostPage.propTypes = {
    startGetPost: PropTypes.func.isRequired,
    startRemovePost: PropTypes.func.isRequired,
    startEditPost: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPost: ( id ) => dispatch( startGetPost( id ) ),
    startEditPost: ( id, updates ) => dispatch( startEditPost( id, updates ) ),
    startRemovePost: ( id ) => dispatch( startRemovePost( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( EditPostPage );