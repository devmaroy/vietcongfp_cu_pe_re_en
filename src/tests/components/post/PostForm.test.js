import React from 'react';
import { shallow } from 'enzyme';
import { PostForm } from '../../../components/post/PostForm';
import TextFieldGroup from '../../../components/common/TextFieldGroup';
import TextAreaFieldGroup from '../../../components/common/TextAreaFieldGroup';
import posts from '../../fixtures/posts';

test( 'should correctly render form inside PostForm', () => {
    const wrapper = shallow( 
        <PostForm 
            user={ { name: 'Josh' } }
            onSubmit={ () => {} }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( 'form' ).exists() ).toBeTruthy();
});

test( 'should correctly render error object if no data provided', () => {
    const wrapper = shallow(
        <PostForm
            user={ { name: 'Josh' } }
            onSubmit={ () => {} }
        />
    );

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });

    expect( wrapper.state( 'errors' ).title.length ).toBeGreaterThan( 0 );
    expect( wrapper.state( 'errors' ).body.length ).toBeGreaterThan( 0 );
});

test( 'should set title on change', () => {
    const value = 'title';

    const wrapper = shallow( 
        <PostForm
            user={ { name: 'Josh' } }
            onSubmit={ () => {} }
        />
    );

    const event = { target: { name: 'title', value } };

    wrapper.find( TextFieldGroup ).simulate( 'change', event );

    expect( wrapper.state( 'title' ) ).toBe( value );
});

test( 'should set body on change', () => {
    const value = 'some message here hiii';

    const wrapper = shallow( 
        <PostForm
            user={ { name: 'Josh' } }
            onSubmit={ () => {} }
        />
    );

    const event = { target: { name: 'body', value } };

    wrapper.find( TextAreaFieldGroup ).simulate( 'change', event );

    expect( wrapper.state( 'body' ) ).toBe( value );
});

test( 'should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();

    const wrapper = shallow(
        <PostForm 
            user={ { name: 'JoshTh' } }
            post={ posts[0] }
            onSubmit={ onSubmitSpy }
        />
    );

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });

    expect( wrapper.state( 'errors' ) ).toEqual( {} );

    expect( onSubmitSpy ).toHaveBeenLastCalledWith({
        title: posts[0].title,
        body: posts[0].body,
        name: posts[0].name,
        createdAt: posts[0].createdAt
    });
});