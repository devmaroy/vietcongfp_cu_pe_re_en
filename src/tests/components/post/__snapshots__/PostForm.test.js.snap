// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render form inside PostForm 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item input-group__item--desktop"
    >
      <TextFieldGroup
        className="text-input"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="input-group"
  >
    <div
      className="input-group__item input-group__item--desktop"
    >
      <TextAreaFieldGroup
        className="textarea textarea--desktop"
        name="body"
        onChange={[Function]}
        placeholder="Text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-meta"
  >
    <button
      className="button"
    >
      Save Post
    </button>
  </div>
</form>
`;
